roxie/ccd/ccdactivities.cpp

```diff
@@ -2050,176 +1718,35 @@ class CRoxieDiskCountActivityFactory : public CRoxieDiskBaseActivityFactory
 
 class CountRecordProcessor : public RecordProcessor
 {
 public:
-    CountRecordProcessor(IInMemoryIndexCursor *_cursor, CRoxieDiskCountActivity &_owner) 
-        : RecordProcessor(_cursor), owner(_owner)
+    CountRecordProcessor(CRoxieDiskCountActivity &_owner, IDirectReader *_reader)
+    : RecordProcessor(_reader), owner(_owner)
     {
         helper = _owner.helper;
     }
-protected:
-    CRoxieDiskCountActivity &owner;
-    IHThorDiskCountArg *helper;
-};
-
-// Used when we have an in-memory key that matches at least some of the filter conditions
-class KeyedCountRecordProcessor : public CountRecordProcessor
-{
-public:
-    KeyedCountRecordProcessor(IInMemoryIndexCursor *_cursor, CRoxieDiskCountActivity &_owner) : CountRecordProcessor(_cursor, _owner)
-    {
-        helper->setCallback(cursor);
-    }
 
-    virtual void doQuery(IMessagePacker *output, unsigned processed, unsigned __int64 rowLimit, unsigned __int64 stopAfter)
-    {
-        // doQuery needs to be as fast as possible - we are making a virtual call to it per query in order to avoid tests of loop-invariants within it
-        unsigned recordSize = owner.meta.getFixedSize();
-        void *recBuffer = output->getBuffer(recordSize, false);
-        IInMemoryIndexCursor *lc = cursor;
-        lc->reset();
-        unsigned __int64 totalCount = 0;
-        while (!aborted)
-        {
-            const void *nextCandidate = lc->nextMatch();
-            if (!nextCandidate)
-                break;
-            totalCount += helper->numValid(nextCandidate);
-            if (totalCount >= stopAfter)
-            {
-                totalCount = stopAfter;
-                break;
-            }
-        }
-
-        if (!aborted)
-        {
-            assert(!owner.serializer); // A count can never need serializing, surely!
-            if (recordSize == 1)
-                *(byte *)recBuffer = (byte)totalCount;
-            else
-            {
-                assertex(recordSize == sizeof(unsigned __int64));
-                *(unsigned __int64 *)recBuffer = totalCount;
-            }
-            output->putBuffer(recBuffer, recordSize, false);
-        }
-    }
-};
-
-IInMemoryFileProcessor *createKeyedCountRecordProcessor(IInMemoryIndexCursor *cursor, CRoxieDiskCountActivity &owner)
-{
-    return new KeyedCountRecordProcessor(cursor, owner);
-}
-
-// Used when there is no key, fixed records
-class UnkeyedCountRecordProcessor : public CountRecordProcessor
-{
-protected:
-    Owned<IDirectReader> reader;
-
-public:
-    UnkeyedCountRecordProcessor(IInMemoryIndexCursor *_cursor, CRoxieDiskCountActivity &_owner, IDirectReader *_reader) 
-        : CountRecordProcessor(_cursor, _owner), reader(_reader)
-    {
-    }
-
-    // This version is used for fixed size rows only - variable size rows use more derived class which overrides
     virtual void doQuery(IMessagePacker *output, unsigned processed, unsigned __int64 rowLimit, unsigned __int64 stopAfter)
```

Comment:
Future: There is potential for a more efficient version of this on keyed data with no post filter - by jumping to the following value, and subtracting the positions.

Reply:
Yes. The old code did not do that either (but could have). Worth raising a Jira